### PR TITLE
DDP-6338 support check_each_instance in workflows

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

See DDP-6338. This PR adds support for `check_each_instance` for activity state workflows. No studies is using this feature right now, so everything should work the same way. Will be updating PanCan study configs to use this in another PR.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

- [x] :relaxed: All good, business as usual!

## How do we demo these changes?

Can tweak PanCan study configs to try it out. Or wait until I update PanCan study configs.

## Testing

- [x] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!

### CODE REVIEW

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/WorkflowDao.java [hunk 1]
```diff
@@ -93,7 +87,8 @@ default long findWorkflowStateIdOrInsert(WorkflowState state) {
             long id = getJdbiWorkflowState().insert(state.getType());
             if (state.getType() == StateType.ACTIVITY) {
                 long activityId = ((ActivityState) state).getActivityId();
-                if (1 != getJdbiWorkflowActivityState().insert(id, activityId)) {
+                boolean checkEachInstance = ((ActivityState) state).shouldCheckEachInstance();
+                if (1 != getJdbiWorkflowState().insertActivityState(id, activityId, checkEachInstance)) {
```
Comment (evgeniipr) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/WorkflowDao.java:92]: How to interpret this "magic number"? Does it make sense to define it as a constant with self-explaining name?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/pex/TreeWalkInterpreter.java [hunk 2]
```diff
@@ -345,7 +347,23 @@ private Object evalFormInstanceQuery(InterpreterContext ictx, PexParser.FormInst
     }
 
     private Object applyFormInstancePredicate(InterpreterContext ictx, PexParser.FormInstancePredicateContext predCtx, long instanceId) {
-        if (predCtx instanceof PexParser.InstanceSnapshotSubstitutionQueryContext) {
+        if (predCtx instanceof PexParser.IsInstanceStatusPredicateContext) {
+            List<InstanceStatusType> expectedStatuses = ((PexParser.IsInstanceStatusPredicateContext) predCtx).STR().stream()
+                    .map(node -> {
```
Comment (evgeniipr) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/pex/TreeWalkInterpreter.java]: The same code is used in the `applyFormPredicate`...
Does it make sense to extract this block inside map() to a separate method:
```
private InstanceStatusType detectInstanceStatusType(String node) {
    String str = extractString(node).toUpperCase();
                        try {
                            return InstanceStatusType.valueOf(str);
                        } catch (Exception e) {
                            throw new PexUnsupportedException("Invalid status used for status predicate: " + str, e);
                        }
}
```
And then use like: `.map(this::detectInstanceStatusType)`

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/WorkflowBuilder.java [hunk 3]
```diff
@@ -91,7 +93,8 @@ private String stateAsStr(Config stateCfg) {
         String type = stateCfg.getString("type");
         if (StateType.ACTIVITY.name().equals(type)) {
             String activityCode = stateCfg.getString("activityCode");
-            return String.format("%s/%s", type, activityCode);
+            boolean checkEachInstance = ConfigUtil.getBoolOrElse(stateCfg, "checkEachInstance", false);
+            return String.format("%s/%s/%s", type, activityCode, checkEachInstance);
```
Comment (evgeniipr) [inline study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/WorkflowBuilder.java]: May be for boolean (checkEachInstance) better to use `%b` ?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/WorkflowDao.java (hunk 1 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/WorkflowDao.java:92]: The `1` here just means did we insert one row... I don't think we need a constant for this.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/pex/TreeWalkInterpreter.java (hunk 2 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/pex/TreeWalkInterpreter.java]: Done

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/WorkflowBuilder.java (hunk 3 above)
Comment (yufengwng) [reply study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/WorkflowBuilder.java]: Done

**Comment by yufengwng** [comment]
@ssettipalli This is ready for review.

**Review by evgeniipr — APPROVED** [review]

**Pull request merged**